### ISSUE TITLE
Output logits differ significantly for different attn_implementations on image inputs

### ISSUE THREAD

**Issue by zzigak** [huggingface/transformers#39067]
### System Info
Hi everyone,
I've been comparing fa2, sdpa, and eager attention implementations. My understand is that these should be very close in logits.

For textual inputs, the relative mean difference is ~1%. For **image** inputs, the relative mean difference is 25% and relative max difference is 75%. These are very large. I attempted this on:

Qwen 2.5 VL 7B, 3B and qwen 2 VL 7B. **See the script below**

Any thoughts?

If you see an error in my script causing this issue, all the better! Thanks.

Thanks!

My env pacakges:

```
  Name                      Version       Build               Channel    
───────────────────────────────────────────────────────────────────────────
  GitPython                 3.1.44        pypi_0              pypi       
  Jinja2                    3.1.4         pypi_0              pypi       
  MarkupSafe                2.1.5         pypi_0              pypi       
  PyYAML                    6.0.2         pypi_0              pypi       
  _libgcc_mutex             0.1           conda_forge         conda-forge
  _openmp_mutex             4.5           2_gnu               conda-forge
  accelerate                1.8.1         pypi_0              pypi       
  aiohappyeyeballs          2.6.1         pypi_0              pypi       
  aiohttp                   3.11.18       pypi_0              pypi       
  aiosignal                 1.3.2         pypi_0              pypi       
  annotated-types           0.7.0         pypi_0              pypi       
  asttokens                 3.0.0         pyhd8ed1ab_1        conda-forge
  attrs                     25.3.0        pypi_0              pypi       
  av                        14.3.0        pypi_0              pypi       
  bitsandbytes              0.45.5        pypi_0              pypi       
  bzip2                     1.0.8         h4bc722e_7          conda-forge
  ca-certificates           2025.4.26     hbd8a1cb_0          conda-forge
  certifi                   2025.4.26     pypi_0              pypi       
  charset-normalizer        3.4.1         pypi_0              pypi       
  click                     8.1.8         pypi_0              pypi       
  comm                      0.2.2         pyhd8ed1ab_1        conda-forge
  datasets                  3.5.1         pypi_0              pypi       
  debugpy                   1.8.14        py311hfdbb021_0     conda-forge
  decorator                 5.2.1         pyhd8ed1ab_0        conda-forge
  decord                    0.6.0         pypi_0              pypi       
  deepspeed                 0.16.7        pypi_0              pypi       
  dill                      0.3.8         pypi_0              pypi       
  docker-pycreds            0.4.0         pypi_0              pypi       
  einops                    0.8.1         pypi_0              pypi       
  exceptiongroup            1.2.2         pyhd8ed1ab_1        conda-forge
  executing                 2.2.0         pyhd8ed1ab_0        conda-forge
  filelock                  3.13.1        pypi_0              pypi       
  frozenlist                1.6.0         pypi_0              pypi       
  fsspec                    2024.6.1      pypi_0              pypi       
  gitdb                     4.0.12        pypi_0              pypi       
  hjson                     3.1.0         pypi_0              pypi       
  huggingface-hub           0.30.2        pypi_0              pypi       
  idna                      3.10          pypi_0              pypi       
  importlib-metadata        8.6.1         pyha770c72_0        conda-forge
  ipykernel                 6.29.5        pyh3099207_0        conda-forge
  ipython                   9.2.0         pyhfb0248b_0        conda-forge
  ipython_pygments_lexers   1.1.1         pyhd8ed1ab_0        conda-forge
  ipywidgets                8.1.6         pypi_0              pypi       
  jedi                      0.19.2        pyhd8ed1ab_1        conda-forge
  jupyter_client            8.6.3         pyhd8ed1ab_1        conda-forge
  jupyter_core              5.7.2         pyh31011fe_1        conda-forge
  jupyterlab_widgets        3.0.14        pypi_0              pypi       
  keyutils                  1.6.1         h166bdaf_0          conda-forge
  krb5                      1.21.3        h659f571_0          conda-forge
  ld_impl_linux-64          2.43          h712a8e2_4          conda-forge
  libedit                   3.1.20250104  pl5321h7949ede_0    conda-forge
  libexpat                  2.7.0         h5888daf_0          conda-forge
  libffi                    3.4.6         h2dba641_1          conda-forge
  libgcc                    14.2.0        h767d61c_2          conda-forge
  libgcc-ng                 14.2.0        h69a702a_2          conda-forge
  libgomp                   14.2.0        h767d61c_2          conda-forge
  liblzma                   5.8.1         hb9d3cd8_0          conda-forge
  libnsl                    2.0.1         hd590300_0          conda-forge
  libsodium                 1.0.20        h4ab18f5_0          conda-forge
  libsqlite                 3.49.1        hee588c1_2          conda-forge
  libstdcxx                 14.2.0        h8f9b012_2          conda-forge
  libstdcxx-ng              14.2.0        h4852527_2          conda-forge
  libuuid                   2.38.1        h0b41bf4_0          conda-forge
  libxcrypt                 4.4.36        hd590300_1          conda-forge
  libzlib                   1.3.1         hb9d3cd8_2          conda-forge
  liger_kernel              0.5.8         pypi_0              pypi       
  markdown-it-py            3.0.0         pypi_0              pypi       
  matplotlib-inline         0.1.7         pyhd8ed1ab_1        conda-forge
  mdurl                     0.1.2         pypi_0              pypi       
  mpmath                    1.3.0         pypi_0              pypi       
  msgpack                   1.1.0         pypi_0              pypi       
  multidict                 6.4.3         pypi_0              pypi       
  multiprocess              0.70.16       pypi_0              pypi       
  ncurses                   6.5           h2d0b736_3          conda-forge
  nest-asyncio              1.6.0         pyhd8ed1ab_1        conda-forge
  networkx                  3.3           pypi_0              pypi       
  ninja                     1.11.1.4      pypi_0              pypi       
  numpy                     2.1.2         pypi_0              pypi       
  nvidia-cublas-cu12        12.4.5.8      pypi_0              pypi       
  nvidia-cuda-cupti-cu12    12.4.127      pypi_0              pypi       
  nvidia-cuda-nvrtc-cu12    12.4.127      pypi_0              pypi       
  nvidia-cuda-runtime-cu12  12.4.127      pypi_0              pypi       
  nvidia-cudnn-cu12         9.1.0.70      pypi_0              pypi       
  nvidia-cufft-cu12         11.2.1.3      pypi_0              pypi       
  nvidia-curand-cu12        10.3.5.147    pypi_0              pypi       
  nvidia-cusolver-cu12      11.6.1.9      pypi_0              pypi       
  nvidia-cusparse-cu12      12.3.1.170    pypi_0              pypi       
  nvidia-cusparselt-cu12    0.6.2         pypi_0              pypi       
  nvidia-ml-py              12.570.86     pypi_0              pypi       
  nvidia-nccl-cu12          2.21.5        pypi_0              pypi       
  nvidia-nvjitlink-cu12     12.4.127      pypi_0              pypi       
  nvidia-nvtx-cu12          12.4.127      pypi_0              pypi       
  opencv-python             4.11.0.86     pypi_0              pypi       
  openssl                   3.5.0         h7b32b05_0          conda-forge
  packaging                 25.0          pyh29332c3_1        conda-forge
  pandas                    2.2.3         pypi_0              pypi       
  parso                     0.8.4         pyhd8ed1ab_1        conda-forge
  peft                      0.15.2        pypi_0              pypi       
  pexpect                   4.9.0         pyhd8ed1ab_1        conda-forge
  pickleshare               0.7.5         pyhd8ed1ab_1004     conda-forge
  pillow                    11.0.0        pypi_0              pypi       
  pip                       25.1.1        pyh8b19718_0        conda-forge
  platformdirs              4.3.7         pyh29332c3_0        conda-forge
  prompt-toolkit            3.0.51        pyha770c72_0        conda-forge
  propcache                 0.3.1         pypi_0              pypi       
  protobuf                  6.30.2        pypi_0              pypi       
  psutil                    7.0.0         py311h9ecbd09_0     conda-forge
  ptyprocess                0.7.0         pyhd8ed1ab_1        conda-forge
  pure_eval                 0.2.3         pyhd8ed1ab_1        conda-forge
  py-cpuinfo                9.0.0         pypi_0              pypi       
  pyarrow                   20.0.0        pypi_0              pypi       
  pydantic                  2.11.3        pypi_0              pypi       
  pydantic_core             2.33.1        pypi_0              pypi       
  pygments                  2.19.1        pyhd8ed1ab_0        conda-forge
  python                    3.11.12       h9e4cc4f_0_cpython  conda-forge
  python-dateutil           2.9.0.post0   pyhff2d567_1        conda-forge
  python_abi                3.11          7_cp311             conda-forge
  pytz                      2025.2        pypi_0              pypi       
  pyzmq                     26.4.0        py311h7deb3e3_0     conda-forge
  qwen-vl-utils             0.0.11        pypi_0              pypi       
  readline                  8.2           h8c095d6_2          conda-forge                                                                                           (main)
  regex                     2024.11.6     pypi_0              pypi       
  requests                  2.32.3        pypi_0              pypi       
  rich                      14.0.0        pypi_0              pypi       
  safetensors               0.5.3         pypi_0              pypi       
  sentry-sdk                2.27.0        pypi_0              pypi       
  setproctitle              1.3.5         pypi_0              pypi       
  setuptools                79.0.1        pyhff2d567_0        conda-forge
  six                       1.17.0        pyhd8ed1ab_0        conda-forge
  smmap                     5.0.2         pypi_0              pypi       
  stack_data                0.6.3         pyhd8ed1ab_1        conda-forge
  sympy                     1.13.1        pypi_0              pypi       
  tensorboardX              2.6.2.2       pypi_0              pypi       
  tk                        8.6.13        noxft_h4845f30_101  conda-forge
  torch                     2.6.0         pypi_0              pypi       
  torchaudio                2.6.0         pypi_0              pypi       
  torchvision               0.21.0        pypi_0              pypi       
  tornado                   6.4.2         py311h9ecbd09_0     conda-forge
  tqdm                      4.67.1        pypi_0              pypi       
  traitlets                 5.14.3        pyhd8ed1ab_1        conda-forge
  transformers              4.53.0        pypi_0              pypi       
  triton                    3.2.0         pypi_0              pypi       
  trl                       0.17.0        pypi_0              pypi       
  typing-inspection         0.4.0         pypi_0              pypi       
  typing_extensions         4.13.2        pyh29332c3_0        conda-forge
  tzdata                    2025b         h78e105d_0          conda-forge
  ujson                     5.10.0        pypi_0              pypi       
  urllib3                   2.4.0         pypi_0              pypi       
  wandb                     0.19.10       pypi_0              pypi       
  wcwidth                   0.2.13        pyhd8ed1ab_1        conda-forge
  wheel                     0.45.1        pyhd8ed1ab_1        conda-forge
  widgetsnbextension        4.0.14        pypi_0              pypi       
  xxhash                    3.5.0         pypi_0              pypi       
  yarl                      1.20.0        pypi_0              pypi       
  zeromq                    4.3.5         h3b0a872_7          conda-forge
  zipp                      3.21.0        pyhd8ed1ab_1        conda-forge
```

### Information

- [x] The official example scripts
- [x] My own modified scripts

### Tasks

- [x] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

Run below for input being text / image (slightly modified script from the tutorial https://huggingface.co/Qwen/Qwen2.5-VL-7B-Instruct). With the newest version of transformers library I was getting an error saying "is_causal is not a property of the model", so had to downgrade. I have the same problem when installing from source.

```py
import torch
from transformers import Qwen2_5_VLForConditionalGeneration, AutoProcessor
from qwen_vl_utils import process_vision_info

torch.manual_seed(42)

processor = AutoProcessor.from_pretrained("Qwen/Qwen2.5-VL-7B-Instruct")
messages = [{
    "role": "user",
    "content": [
        # {"type": "image", "image": "https://qianwen-res.oss-cn-beijing.aliyuncs.com/Qwen-VL/assets/demo.jpeg"},
        {"type": "text", "text": "Describe this simage."}
    ],
}]
text = processor.apply_chat_template(messages, tokenize=False, add_generation_prompt=True)
image_inputs, video_inputs = process_vision_info(messages)

inputs = processor(
    text=[text],
    images=image_inputs,
    videos=video_inputs,
    padding=True,
    return_tensors="pt"
)

model_flash = Qwen2_5_VLForConditionalGeneration.from_pretrained(
    "Qwen/Qwen2.5-VL-3B-Instruct",
    torch_dtype=torch.bfloat16,
    attn_implementation="sdpa",
    device_map={"": 0},
)
model_flash.eval()

model_sdpa = Qwen2_5_VLForConditionalGeneration.from_pretrained(
    "Qwen/Qwen2.5-VL-3B-Instruct",
    torch_dtype=torch.bfloat16,
    attn_implementation="flash_attention_2",  # or omit
    # attn_implementation="eager",
    device_map={"": 1},
)
model_sdpa.eval()

# === Run FlashAttention2 ===
with torch.no_grad():
    inputs_flash = {k: v.to("cuda:0") for k, v in inputs.items()}
    out_flash = model_flash(**inputs_flash).logits.cpu()

# === Run SDPA ===
with torch.no_grad():
    inputs_sdpa = {k: v.to("cuda:1") for k, v in inputs.items()}
    out_sdpa = model_sdpa(**inputs_sdpa).logits.cpu()

diff = (out_flash - out_sdpa).abs()
print(f"Max abs diff: {diff.max().item():.6f}")
print(f"Mean abs diff: {diff.mean().item():.6f}")
print(f"Relative max diff: {(diff.max() / out_flash.abs().max()).item():.6f}")
print(f"Relative mean diff: {(diff.mean() / out_flash.abs().mean()).item():.6f}")

```

### Expected behavior

Relative error for logits on image inputs being comparable to text inputs and much lower than 25%.

**Comment by ArthurZucker** [comment]
cc @vasqu if you have time to take a look! Might just be attention mask related, or expected

**Comment by zucchini-nlp** [comment]
@zzigak hey!

The issue in Qwen vision module lies in how attention is implemented for FA2 specifically. The model uses `fa2_varlen` for ragged inputs thus computing attention for small chunks of vision states separately. While SDPA/eager attentions try to match it by creating a mutually exclusive attention mask for each chunk like below, that isn't in fact same as if we did a padded attention because the softmax would be applied along the seq-length dim

https://github.com/huggingface/transformers/blob/ea0ea392e57f8816f9ab8e5f740577a0343a1594/src/transformers/models/qwen2_vl/modeling_qwen2_vl.py#L354-L361

Right now there's a PR from Qwen team (https://github.com/huggingface/transformers/pull/37363) which seems to switch to a loop for computing attention, which I believe would match with `allclose` to FA2, though I did not test it yet. I will first merge a fix for FA2 with `module has no attribute "is_causal"` and then test out logits with a looped or padded attention

**Comment by zzigak** [comment]
Hi @zucchini-nlp, thank you so much for looking into this! 
Should I wait a bit and pull the newest version of transformers once the issue is fixed / PR from the Qwen team is accepted?
Also, if I'm understanding this correctly, at the current state, non of the attentions (SDPA, FA2, eager) produce "correct" logits? Or is the FA2 correct, but SDPA / eager are trying to match it incorrectly?

Thanks again! 
Looking forward to the fix

**Comment by zucchini-nlp** [comment]
I believe the FA2 is the "correct" version as the original implementation of the model uses FA2 first. We added other implementations so it is still supported though it's not a perfect match in logits

Let's wait for the PRs to be merged, yes :)

**Comment by JJJYmmm** [comment]
Hi @zucchini-nlp @zzigak,
I suspect the logits diff is due to the Qwen-ViT's numerical sensitivity(in fact, the model is very sensitive to input image size), not a code issue. 🧐

Two observations:
1. The relative mean diff varies significantly (from 2% to 20%) just by changing the input image resolution (max_pixels).
2. The diff clearly accumulates layer by layer. The initial error from the ViT's own attention layers is amplified as it propagates.

Here’s a quick look at the error accumulation(This experiment is based on #37363, brings the logits diff down from 25%~ to 19%~):
```python
messages = [{
    "role": "user",
    "content": [
        {"type": "image", "image": "https://qianwen-res.oss-cn-beijing.aliyuncs.com/Qwen-VL/assets/demo.jpeg", "max_pixels": 10 * 28 * 28},
        {"type": "text", "text": "Describe this simage."}
    ],
}]

# max_pixels = 10 * 28 * 28
layer 0 Relative mean diff: 0.016968
layer 5 Relative mean diff: 0.009277
layer 10 Relative mean diff: 0.010376
layer 15 Relative mean diff: 0.010864
layer 20 Relative mean diff: 0.011719
layer 25 Relative mean diff: 0.012634
layer 30 Relative mean diff: 0.013428
layer 31 Relative mean diff: 0.013428
layer 32 Relative mean diff: 0.013855
layer 33 Relative mean diff: 0.014465
layer 34 Relative mean diff: 0.015137
layer 35 Relative mean diff: 0.015991
layer 36 Relative mean diff: 0.020264

# max_pixels = 2000*28*28
layer 0 Relative mean diff: 0.043945
layer 5 Relative mean diff: 0.045410
layer 10 Relative mean diff: 0.053711
layer 15 Relative mean diff: 0.067383
layer 20 Relative mean diff: 0.084961
layer 25 Relative mean diff: 0.108398
layer 30 Relative mean diff: 0.135742
layer 31 Relative mean diff: 0.142578
layer 32 Relative mean diff: 0.149414
layer 33 Relative mean diff: 0.156250
layer 34 Relative mean diff: 0.162109
layer 35 Relative mean diff: 0.164062
layer 36 Relative mean diff: 0.159180
```

**Comment by zzigak** [comment]
@zucchini-nlp Any updates on this?

**Comment by zucchini-nlp** [comment]
It's the issue with Qwen-ViT numberical sensitivity as commented above, so what we can do is to try to use higher fp precision to get better logit match. There's no bug in code as per comment from Qwen team

> I suspect the logits diff is due to the Qwen-ViT's numerical sensitivity(in fact, the model is very sensitive to input image size), not a code issue. 🧐
